File: Python/day22/AOC.py
from collections import defaultdict

Prices = {}
def get_secret_number(secret: int) -> int:
    key = secret
    Prices[key] = []
    for _ in range(2000):
        secret ^= (secret * 64) % 16777216
        secret ^= (secret // 32) % 16777216
        secret ^= (secret * 2048) % 16777216
        if key in Prices: # we need this for part 2, so we'll calculate these in part 1
            Prices[key].append(secret%10)
    return secret

def part1(input: list[str]) -> int:
    return sum([get_secret_number(initial_secret) for initial_secret in [int(line) for line in input]])

def get_sequences(price: list[int], changes: list[int]) -> dict:
    sequence = defaultdict(int)
    for i in range(len(changes)-3):
        pattern = tuple(changes[i:i+4])
        if pattern not in sequence:
            sequence[pattern] = price[i+4]
            
    return sequence

def part2(input: list[str]) -> int:
    scores = defaultdict(int)
    for line in input:
        if int(line) not in Prices: # should literally only happen during testing, since the example input changes.
            get_secret_number(int(line))
        price = Prices[int(line)]
        changes = [price[i+1] - price[i] for i in range(len(price)-1)]
        seq = get_sequences(price, changes)
        for k,v in seq.items():
            scores[k] += v

    return max(scores.values())

def main():
    input = open("input.txt", "r").read().splitlines()
    print(f"Part 1: {part1(input)}")
    print(f"Part 2: {part2(input)}")

if __name__ == "__main__":
    main()

File: Go/day22/AOC.go
package main

import (
	"fmt"
	"os"
	"strings"
	"utils"
)

func nextNum(x int) int {
	x ^= (x * 64) % 16777216
	x ^= (x / 32) % 16777216
	x ^= (x * 2048) % 16777216
	return x
}

func part1(input []string) int {
	nums := utils.Apply(input, utils.Int)

	return utils.ApplySum(nums, func(x int) int {
		for i := 0; i < 2000; i++ {
			x = nextNum(x)
		}
		return x
	})
}

type Sequence struct{ a, b, c, d int }

func getSequences(price []int, change []int) map[Sequence]int {
	seq := map[Sequence]int{}
	for i := 0; i < len(change)-3; i++ {
		pattern := Sequence{change[i], change[i+1], change[i+2], change[i+3]}
		if _, ok := seq[pattern]; !ok {
			seq[pattern] = price[i+4]
		}
	}
	return seq
}

func part2(input []string) int {
	nums := utils.Apply(input, utils.Int)
	prices := map[int][]int{}
	scores := map[Sequence]int{}
	for _, x := range nums {
		key := x
		prices[key] = make([]int, 2000)
		for i := 0; i < 2000; i++ {
			x = nextNum(x)
			prices[key][i] = x % 10
		}
		price := prices[key]
		change := []int{}
		for i := 0; i < len(price)-1; i++ {
			change = append(change, price[i+1]-price[i])
		}
		seq := getSequences(price, change)
		for k, v := range seq {
			scores[k] += v
		}
	}
	max := 0
	for _, v := range scores {
		if v > max {
			max = v
		}
	}
	return max
}

func main() {
	input, err := os.ReadFile("input.txt")
	if err != nil {
		fmt.Print("Error reading file\n")
		panic(err)
	}

	lines := strings.Split(string(input), "\n")
	part1_ans := part1(lines)
	fmt.Printf("Part 1: %d\n", part1_ans)

	part2_ans := part2(lines)
	fmt.Printf("Part 2: %d\n", part2_ans)
}
